api/v2/api.go

```diff
@@ -389,6 +341,144 @@ func (api *API) postAlertsHandler(params alert_ops.PostAlertsParams) middleware.
 	return alert_ops.NewPostAlertsOK()
 }
 
+func (api *API) getAlertGroupsHandler(params alertgroup_ops.GetAlertGroupsParams) middleware.Responder {
+	var (
+		err            error
+		receiverFilter *regexp.Regexp
+		matchers       = []*labels.Matcher{}
+	)
+
+	for _, matcherString := range params.Filter {
+		matcher, err := parse.Matcher(matcherString)
+		if err != nil {
+			level.Error(api.logger).Log("msg", "failed to parse matchers", "err", err)
+			return alertgroup_ops.NewGetAlertGroupsBadRequest().WithPayload(err.Error())
+		}
+
+		matchers = append(matchers, matcher)
+	}
+
+	if params.Receiver != nil {
+		receiverFilter, err = regexp.Compile("^(?:" + *params.Receiver + ")$")
+		if err != nil {
+			level.Error(api.logger).Log("msg", "failed to compile receiver regex", "err", err)
+			return alertgroup_ops.
+				NewGetAlertGroupsBadRequest().
+				WithPayload(
+					fmt.Sprintf("failed to parse receiver param: %v", err.Error()),
+				)
+		}
+	}
+
+	rf := routeFilter(receiverFilter)
+	af := api.alertFilter(matchers, *params.Silenced, *params.Inhibited, *params.Active)
+	alertGroups, allReceivers := api.alertGroups(rf, af)
+
+	res := make(open_api_models.AlertGroups, 0, len(alertGroups))
+
+	for _, alertGroup := range alertGroups {
+		ag := &open_api_models.AlertGroup{
+			Receiver: &open_api_models.Receiver{Name: &alertGroup.Receiver},
+			Labels:   modelLabelSetToAPILabelSet(alertGroup.Labels),
+			Alerts:   make([]*open_api_models.GettableAlert, 0, len(alertGroup.Alerts)),
+		}
+
+		for _, alert := range alertGroup.Alerts {
+			fp := alert.Fingerprint()
+			receivers := allReceivers[fp]
+			status := api.getAlertStatus(fp)
+			apiAlert := alertToOpenAPIAlert(alert, status, receivers)
+			ag.Alerts = append(ag.Alerts, apiAlert)
+		}
+		res = append(res, ag)
+	}
+
+	return alertgroup_ops.NewGetAlertGroupsOK().WithPayload(res)
+}
+
+func (api *API) alertFilter(matchers []*labels.Matcher, silenced, inhibited, active bool) func(a *types.Alert, now time.Time) bool {
+	return func(a *types.Alert, now time.Time) bool {
+		if !a.EndsAt.IsZero() && a.EndsAt.Before(now) {
+			return false
+		}
+
+		// Set alert's current status based on its label set.
+		api.setAlertStatus(a.Labels)
+
+		// Get alert's current status after seeing if it is suppressed.
+		status := api.getAlertStatus(a.Fingerprint())
+
+		if !active && status.State == types.AlertStateActive {
+			return false
+		}
+
+		if !silenced && len(status.SilencedBy) != 0 {
+			return false
+		}
+
+		if !inhibited && len(status.InhibitedBy) != 0 {
+			return false
+		}
+
+		if !alertMatchesFilterLabels(&a.Alert, matchers) {
+			return false
+		}
+
+		return true
+	}
+}
+
+func routeFilter(receiverFilter *regexp.Regexp) func(r *dispatch.Route) bool {
```

Comment:
(nit) unless you wanted to test it in isolation, I would inline this function in `getAlertGroupsHandler()` since it is only used there.